### PR TITLE
Fixes required for Solang

### PR DESCRIPTION
**Pull request by seanyoung**

### CODE REVIEW

File: clang/lib/Basic/Targets/BPF.h [hunk 1]
```diff
@@ -112,6 +112,8 @@ class LLVM_LIBRARY_VISIBILITY BPFTargetInfo : public TargetInfo {
     StringRef CPUName(Name);
     return isValidCPUName(CPUName);
   }
+
+  bool hasExtIntType() const override { return true; }
```
Comment (dmakarov) [inline clang/lib/Basic/Targets/BPF.h]: maybe this should depend on whether the `solana` feature is enabled? The base class `TargetInfo` method `hasExtIntType` returns `false`.

Comment (seanyoung) [reply clang/lib/Basic/Targets/BPF.h]: Yes, it should. In fact I've tried to upstream this patch and got the feedback that these types need enabling in the linux kernel. I started working on this https://www.spinics.net/lists/netdev/msg711156.html but then I got the feedback "why do we need this in the kernel" for which I didn't have a good answer.

Making this solana-only sidesteps that entire discussion.

File: lld/tools/lld/lld.cpp [hunk 2]
```diff
@@ -136,10 +137,22 @@ static Flavor parseFlavor(std::vector<const char *> &v) {
...
+// LTO calls LLVM functions to compile bitcode files to native code.
+// Technically this can be delayed until we read bitcode files, but
+// we don't bother to do lazily because the initialization is fast.
+static void initLLVM() {
+  InitializeAllTargets();
+  InitializeAllTargetMCs();
+  InitializeAllAsmPrinters();
+  InitializeAllAsmParsers();
+}
+
```
Comment (dmakarov) [inline lld/tools/lld/lld.cpp:150]: Wouldn't this affect the linker for other platforms, COFF, for example, by calling these functions twice, first here and then again in `LinkerDriver::link`

Comment (seanyoung) [reply lld/tools/lld/lld.cpp:150]: Good point. I'll fix that too.

**Comment by dmakarov** [comment]
I add the prefix `[SOL][BPF]` to every commit message, to distinguish visually the commits that are Solana specific, so I don't miss them when I update the source tree to a new snapshot of llvm-project. It would be nice if your commit message had this prefix too (`[BPF]` may be not applicable to clang or lld changes, but I would appreciate having the `[SOL]` prefix).

File: clang/lib/Basic/Targets/BPF.h (hunk 1 above)
Comment (seanyoung) [reply clang/lib/Basic/Targets/BPF.h]: Fixed

File: lld/tools/lld/lld.cpp (hunk 2 above)
Comment (seanyoung) [reply lld/tools/lld/lld.cpp:150]: When I checked the other targets, only coff has these calls -- I've removed them.

**Comment by seanyoung** [comment]
I've updated the commit messages.

**Comment by dmakarov** [comment]
LGTM. Let's wait for the CI checks to finish (it'll take some time), and then this can be merged. Thank you.

**Comment by seanyoung** [comment]
Looks like all CI checks have passed

**Pull request merged**